subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/recomp/GroovyRecompilationSpecProvider.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.recomp;
+
+import org.gradle.api.internal.file.FileOperations;
+import org.gradle.api.internal.file.FileTreeInternal;
+import org.gradle.api.internal.tasks.compile.JavaCompileSpec;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.internal.Factory;
+import org.gradle.internal.execution.history.changes.DefaultFileChange;
+import org.gradle.util.RelativePathUtil;
+import org.gradle.work.FileChange;
+import org.gradle.work.InputChanges;
+
+import java.io.File;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public class GroovyRecompilationSpecProvider extends AbstractRecompilationSpecProvider {
+    private final CompilationSourceDirs sourceDirs;
+    private final InputChanges inputChanges;
+    private final Iterable<FileChange> classpathChanges;
+    private final Iterable<FileChange> sourceChanges;
+    private final SourceFileClassNameConverter sourceFileClassNameConverter;
+
+    public GroovyRecompilationSpecProvider(FileOperations fileOperations,
+                                           FileTreeInternal sources,
+                                           InputChanges inputChanges,
+                                           Iterable<FileChange> sourceChanges,
+                                           Iterable<FileChange> classpathChanges,
+                                           SourceFileClassNameConverter sourceFileClassNameConverter) {
+        super(fileOperations, sources);
+        this.sourceDirs = new CompilationSourceDirs(sources);
+        this.inputChanges = inputChanges;
+        this.classpathChanges = classpathChanges;
```

Comment:
`classpathChanges` is not used anymore in this class. We now re-construct the classpath changes via `classpathSnapshotProvider`.

Reply:
Removed.